### PR TITLE
feat(openshift): add mTLS for Prometheus metrics

### PR DESCRIPTION
**Pull request by jkhelil**
Secure component metrics endpoints with mutual TLS on OpenShift. Each reconciler syncs the Prometheus client CA bundle from kube-system/extension-apiserver-authentication into the component namespace as a metrics-client-ca ConfigMap.

OpenShift's serving-cert controller is triggered via annotation on each metrics Service to provision a per-component TLS Secret. Two new manifest transformers (InjectMetricsServingCert, ApplyMetricsTLS) wire the Secret and ConfigMap as volumes and inject METRICS_PROMETHEUS_TLS_* env vars so that the knative/pkg prometheus.Server enables mTLS with require client auth.

ServiceMonitor resources are updated with scheme: https, scrapeClass tls-client-certificate-auth, and tlsConfig.serverName. The existing UpdateServiceMonitorTargetNamespace transformer is extended to also patch the namespace segment inside serverName at runtime.

Relates-To: SRVKP-8172

Assisted-by: Claude Sonnet 4.6 (via Cursor)

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Run `make test lint` before submitting a PR
- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
Tekton Operator now enables mTLS between Prometheus and all Tekton
component metrics endpoints on OpenShift. Each component namespace
receives a synced client-CA ConfigMap and a serving-cert Secret so
that the knative/pkg prometheus.Server can enforce mutual TLS on its
/metrics endpoint.
```

### CODE REVIEW

File: pkg/reconciler/openshift/tektonresult/extension.go [hunk 1]
```diff
@@ -108,6 +112,15 @@ func (oe *openshiftExtension) Transformers(comp v1alpha1.TektonComponent) []mf.T
 		injectLokiStackTLSCACert(instance.Spec.LokiStackProperties),
 		injectResultsAPIServiceCACert(instance.Spec.ResultsAPIProperties),
 		injectPostgresUpgradeSupport(),
+		// mTLS for Prometheus scraping.
+		// tekton-results-watcher is a StatefulSet that exposes a "metrics" port.
+		// tekton-results-api-service is a Deployment that exposes a "prometheus" port.
+		occommon.InjectMetricsServingCertWithPort(tektonResultWatcherName, "metrics"),
+		occommon.ApplyMetricsTLS("StatefulSet", tektonResultWatcherName,
+			occommon.MetricsServingCertSecretName(tektonResultWatcherName)),
+		occommon.InjectMetricsServingCertWithPort(serviceAPI, "prometheus"),
```
Comment (infernus01) [inline pkg/reconciler/openshift/tektonresult/extension.go]: This might create annotation conflict with [injectResultsAPISeerviceCACert](https://github.com/tektoncd/operator/pull/3473/changes#diff-a08d70d4f7f1a1932aafcdf80de931bc4218ee3838bc09fb6db75139805ba3c5R113) as this already sets [service.beta.openshift.io/serving-cert-secret-name](https://github.com/tektoncd/operator/pull/3473/changes#diff-a08d70d4f7f1a1932aafcdf80de931bc4218ee3838bc09fb6db75139805ba3c5R243) to "tekton-results-tls" on tekton-results-api-service. This transformer overwrites that same annotation with "tekton-results-api-service-metrics-tls" (via [MetricsServingCertSecretName](https://github.com/tektoncd/operator/pull/3473/changes#diff-0efc7d6f2deab46ddecca3fb520c396bb586308ff5c3609005ff85a406898ff4R66)).

File: cmd/openshift/operator/kodata/openshift-monitoring/00-monitoring.yaml [hunk 2]
```diff
@@ -58,8 +58,14 @@ metadata:
 spec:
   endpoints:
     - interval: 10s
-      port: http-metrics
+      port: https-metrics
       honorLabels: true
+      scheme: https
+      tlsConfig:
```
[the remark this answers is not in the document]
Comment (jkhelil) [reply cmd/openshift/operator/kodata/openshift-monitoring/00-monitoring.yaml:64]: scrapeClass is not working on ocp 4.19, we inject directly the client cert

**Review by infernus01 — COMMENTED** [review]
/lgtm

File: pkg/reconciler/openshift/common/metricstls.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+	// Add the client-CA ConfigMap volume.
+	clientCAVol := corev1.Volume{
+		Name: metricsClientCAVolume,
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: MetricsClientCAConfigMap,
+				},
+			},
```
[the remark this answers is not in the document]
Comment (jkhelil) [reply pkg/reconciler/openshift/common/metricstls.go:259]: The ordering guarantee is the fix. EnsureMetricsClientCA runs in TektonConfig.PreReconcile, which executes before any child component (TektonPipeline, TektonResult, etc.) is reconciled and before their pods are created. The ConfigMap is always present by the time pods are scheduled.

If the ConfigMap is deleted after install, TektonConfig's periodic reconciliation recreates it. In the small window between deletion and recreation, a pod restart will block — which is intentional. A pod without the client-CA bundle cannot perform mTLS client verification at all, so failing loudly (pod unschedulable) is preferable to starting silently with broken mTLS. This is consistent with how other mandatory ConfigMaps (e.g. CA bundles) are handled throughout the operator.

File: pkg/reconciler/openshift/common/transformer.go [hunk 4]
```diff
@@ -176,15 +178,67 @@ func UpdateServiceMonitorTargetNamespace(targetNamespace string) mf.Transformer
...
 		return nil
 	}
 }
 
+// replaceServerNameNamespace replaces the namespace segment in a Kubernetes
+// in-cluster DNS name of the form "<service>.<namespace>.svc[…]" with
+// targetNamespace. Returns the original string unchanged if it does not match.
+func replaceServerNameNamespace(serverName, targetNamespace string) string {
+	// Preserve any suffix after ".svc" (e.g. ".cluster.local").
+	const svsSuffix = ".svc"
```
Comment (enarha) [inline pkg/reconciler/openshift/common/transformer.go]: typo? `svsSuffix` -> `svcSuffix`

File: pkg/reconciler/openshift/common/transformer.go [hunk 5]
```diff
@@ -176,15 +178,67 @@ func UpdateServiceMonitorTargetNamespace(targetNamespace string) mf.Transformer
...
+		}
+
 		return nil
 	}
 }
 
+// replaceServerNameNamespace replaces the namespace segment in a Kubernetes
+// in-cluster DNS name of the form "<service>.<namespace>.svc[…]" with
+// targetNamespace. Returns the original string unchanged if it does not match.
+func replaceServerNameNamespace(serverName, targetNamespace string) string {
```
Comment (enarha) [inline pkg/reconciler/openshift/common/transformer.go:226]: replaceServerNameNamespace is easy to test and I consider it worth adding test.

File: pkg/reconciler/openshift/tektonchain/extension.go [hunk 6]
```diff
@@ -25,21 +25,28 @@ import (
 	operatorclient "github.com/tektoncd/operator/pkg/client/injection/client"
 	"github.com/tektoncd/operator/pkg/reconciler/common"
 	occommon "github.com/tektoncd/operator/pkg/reconciler/openshift/common"
+	"k8s.io/client-go/kubernetes"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
 )
 
 const (
 	tektonChainsControllerName = "tekton-chains-controller"
+	// tektonChainsMetricsService is the Service that exposes the http-metrics
```
Comment (enarha) [inline pkg/reconciler/openshift/tektonchain/extension.go]: `exposes the http-metrics` is a bit misleading because after the transformation this will be https. I suggest you entirely drop the http/https reference e.g. "tektonChainsMetricsService is the metrics Service for chains."

File: pkg/reconciler/openshift/openshiftpipelinesascode/extension.go [hunk 7]
```diff
@@ -106,16 +123,14 @@ func (oe *openshiftExtension) Transformers(comp v1alpha1.TektonComponent) []mf.T
...
-			// injecting with no prefix is harmless and keeps them consistent for future use.
 			occommon.InjectTLSEnvVars(oe.resolvedTLSConfig, "Deployment", pacControllerDeployment, []string{pacControllerContainerName}, ""),
 		)
 	}
 
 	return trns
 }
 
-func (oe *openshiftExtension) PreReconcile(ctx context.Context, _ v1alpha1.TektonComponent) error {
+func (oe *openshiftExtension) PreReconcile(ctx context.Context, comp v1alpha1.TektonComponent) error {
```
Comment (enarha) [inline pkg/reconciler/openshift/openshiftpipelinesascode/extension.go]: `comp` is not really used in that function. Why change `_`?

**Review by enarha — COMMENTED** [review]
metricsca.go, metricstls.go were added, but the PR does not include tests for these. Here are few things we can tests (asked Claude to generate):

## Test recommendations for `metricsca.go`, `metricstls.go`, and `UpdateServiceMonitorTargetNamespace`

### `metricsca.go` — `EnsureMetricsClientCA`

Unit tests using `fake.NewSimpleClientset()`:

| Case | What it verifies |
|------|-----------------|
| Source ConfigMap missing | Returns error, no target ConfigMap created |
| Source ConfigMap exists but key `client-ca-file` absent | Returns error with descriptive message |
| Target ConfigMap doesn't exist | Creates it with correct name, namespace, labels, data |
| Target ConfigMap exists, same CA bundle | No update call (verify via fake client action list) |
| Target ConfigMap exists, different CA bundle | Updates with new bundle, preserves other fields |
| Target ConfigMap exists, `Data` is nil | Updates without panic |

These are straightforward fake-clientset tests. The function has clear branch points that map 1:1 to test cas
es. Verify side effects through the fake client's `Actions()` — e.g. confirm no `Update` action fires in the 
no-op case.

### `metricstls.go` — transformer unit tests

Each transformer takes an `*unstructured.Unstructured` and mutates it. The existing `cabundle_test.go` and `t
ransformer_test.go` already establish the pattern: build an unstructured object, run the transformer, assert 
on the result.

**`AnnotateMetricsServingCert`**:
- Matching Service with no annotations → annotation added
- Matching Service with existing annotations → annotation added, others preserved
- Non-matching Service name → unchanged
- Non-Service resource (e.g. Deployment) → unchanged

**`RenameServicePort`**:
- Port matching `fromPort` → renamed to `toPort`
- No port matches `fromPort` → unchanged (no error)
- Multiple ports, only one matches → only that one renamed
- Non-matching Service → unchanged

**`UpdateServiceMonitorForMetricsMTLS`**:
- Matching ServiceMonitor → port renamed, scheme set to `https`, tlsConfig injected with correct serverName
- Non-matching name → unchanged
- No endpoints in spec → unchanged (no error)
- Multiple endpoints → all get scheme + tlsConfig

**`ApplyMetricsTLS`**:
- Deployment: volumes, volume mounts, and env vars injected into all containers
- StatefulSet: same
- Non-matching kind or name → unchanged
- **Idempotency**: run twice, result is identical to running once (this is important — `AddOrReplaceInList` should handle it, but verify)
- Multi-container pod → all containers get mounts and env vars

**`replaceServerNameNamespace`** — pure function, table-driven:

| Input | Expected output |
|-------|-----------------|
| `"svc.openshift-pipelines.svc"` | `"svc.new-ns.svc"` |
| `"svc.openshift-pipelines.svc.cluster.local"` | `"svc.new-ns.svc.cluster.local"` |
| `"no-dot-svc"` | `"no-dot-svc"` (unchanged) |
| `"svc.ns.nomatch"` | `"svc.ns.nomatch"` (unchanged) |
| `""` | `""` (unchanged) |

### `UpdateServiceMonitorTargetNamespace` (refactored in `transformer.go`)

The refactored function now does two things: replaces `matchNames[0]` and patches `tlsConfig.serverName`. Test them together since they're one transformer:

| Case | What it verifies |
|------|-----------------|
| ServiceMonitor with matchNames + tlsConfig.serverName | Both namespace segments replaced |
| ServiceMonitor with matchNames, no tlsConfig | matchNames updated, no error on missing tlsConfig |
| ServiceMonitor with no matchNames | No panic (old code would panic on this) |
| ServiceMonitor with multiple endpoints, only some have tlsConfig | Only endpoints with tlsConfig get serverName patched |
| Non-ServiceMonitor resource | Unchanged |

The regression value here is high — the old code had bare type assertions that would panic on unexpected structure. The new tests should confirm that malformed inputs produce errors, not panics.

### Priority

Highest-value tests to add first:

1. `replaceServerNameNamespace` — pure function, quick to write, catches real edge cases
2. `ApplyMetricsTLS` idempotency — the most complex transformer, most likely to regress
3. `EnsureMetricsClientCA` create-vs-update — branching logic with real operational consequences

**Review by enarha — COMMENTED** [review]
/approve

**Review by infernus01 — COMMENTED** [review]
/lgtm
